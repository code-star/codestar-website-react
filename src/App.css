.text-center {
	text-align: center;
}

p {
	color: white;
}

.mh-full {
	min-height: 100vh;
}
