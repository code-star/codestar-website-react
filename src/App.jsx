import React, { Component } from 'react';
import AsyncComponent from './AsyncComponent/AsyncComponent';
import { BrowserRouter as Router, Route } from 'react-router-dom';
import CssBaseline from 'material-ui/CssBaseline';

import NavBar from './NavBar/NavBar';
import SideMenu from './SideMenu/SideMenu';
import Footer from './Footer/Footer';

import casesList from './Cases/CasesList';

const AsyncIntro = AsyncComponent(() => import('./Intro/Intro'));
const AsyncCases = AsyncComponent(() => import('./Cases/Cases'));
const AsyncCaseDetails = AsyncComponent(() =>
	import('./CaseDetails/CaseDetails')
);
const AsyncAbout = AsyncComponent(() => import('./About/About'));
const AsyncJobs = AsyncComponent(() => import('./Jobs/Jobs'));
const AsyncContact = AsyncComponent(() => import('./Contact/Contact'));

class App extends Component {
	state = {
		drawerMenu: false,
	};

	constructor(props) {
		super();
		this.classes = props.classes;
	}

	toggleDrawer = () => {
		this.setState(prevState => {
			return {
				drawerMenu: !prevState.drawerMenu,
			};
		});
	};

	render() {
		return (
			<Router>
				<div>
					<CssBaseline />
					<NavBar toggle={this.toggleDrawer} />
					<SideMenu open={this.state.drawerMenu} toggle={this.toggleDrawer} />

					<Route exact path="/" component={AsyncIntro} />
					<Route exact path="/cases" component={AsyncCases} />

					{casesList.map(clientCase => (
						<Route
							exact
							path={`/cases/${clientCase.path}`}
							key={clientCase.client}
							render={() => <AsyncCaseDetails {...clientCase} />}
						/>
					))}

					<Route path="/about" component={AsyncAbout} />
					<Route path="/jobs" component={AsyncJobs} />
					<Route path="/contact" component={AsyncContact} />

					<Footer />
				</div>
			</Router>
		);
	}
}

export default App;
