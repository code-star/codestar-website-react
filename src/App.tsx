import React, { Component, lazy, Suspense } from 'react';
import { Redirect, Route, Router, Switch } from 'react-router-dom';
import createHistory from 'history/createBrowserHistory';
import { History } from 'history';
import { CssBaseline } from '@material-ui/core';
import { MuiThemeProvider } from '@material-ui/core/styles';
import theme from './codestarMuiTheme';
import ScrollToTop from './ScrollToTop';
import Footer from './Footer/Footer';
// import jobsList from './Jobs/JobsList';
import NavContainer from './containers/NavContainer/NavContainer';
// import registerServiceWorker, { onRegistration } from './registerServiceWorker';
import AppMessageSnackbar from './components/Molecules/AppMessageSnackbar/AppMessageSnackbar';
import LoadingMessage from './components/Atoms/LoadingMessage/LoadingMessage';
import FullHeight from './components/Atoms/FullHeight/FullHeight';

// const AsyncIntro = lazy(() => import('./Intro/Intro'));
// const AsyncCases = lazy(() => import('./Cases/Cases'));
// const AsyncAbout = lazy(() => import('./components/Pages/About/About'));
// const AsyncJobs = lazy(() => import('./Jobs/Jobs'));
// FIXME any type, lazy<JobDescriptionOuterProps> is invalid
// const AsyncJobDescription = lazy<any>(() =>
//   import('./JobDescription/JobDescription')
// );
// const AsyncContact = lazy(() => import('./Contact/Contact'));
const AsyncCodeChallenge = lazy(() =>
  import('./components/Pages/CodeChallenge/CodeChallenge')
);
const AsyncNotFound = lazy(() => import('./NotFound/NotFound'));
const AsyncEvents = lazy(() =>
  import('./containers/EventsContainer/EventsContainer')
);
const AsyncPublications = lazy(() =>
  import('./components/Pages/Publications/Publications')
);

// Darkest -> lightest background
const pages = [
  '',
  'publications',
  'events',
  'codelancer',
  // 'cases',
  // 'about',
  // 'jobs',
  // 'contact',
];

type AppProps = Readonly<{}>;
type AppState = Readonly<{
  showMessage: boolean;
  message: string | null;
}>;

class App extends Component<AppProps, AppState> {
  private history: History;

  constructor(props: AppProps) {
    super(props);
    this.history = createHistory({ basename: process.env.PUBLIC_URL });

    this.history.listen(location =>
      this.updateBackgroundColor(location.pathname)
    );
    this.updateBackgroundColor(this.history.location.pathname);

    this.state = {
      showMessage: false,
      message: null,
    };

    // The serviceworker seems to conflict with subdomains like https://code-star.github.io/nx-reference/ 
    // That URL would always be redirected to the 404 page.
//     const serviceWorkerRegistration: Promise<ServiceWorkerRegistration | void> = registerServiceWorker();
//     serviceWorkerRegistration
//       .then(
//         onRegistration(message => {
//           this.setState({
//             showMessage: true,
//             message,
//           });
//         })
//       )
//       .catch(error => {
//         console.error('Error during service worker registration:', error);
//       });
  }

  handleCloseSnackbar = () => {
    this.setState({ showMessage: false });
  };

  handleExitedSnackbar = () => {
    this.setState({ message: null });
  };

  public render() {
    return (
      <Router history={this.history}>
        <MuiThemeProvider theme={theme}>
          <CssBaseline />

          <AppMessageSnackbar
            showMessage={this.state.showMessage}
            message={this.state.message}
            handleCloseSnackbar={this.handleCloseSnackbar}
            handleExitedSnackbar={this.handleExitedSnackbar}
          />

          <NavContainer history={this.history} />

          <FullHeight>
            <ScrollToTop>
              <Suspense fallback={<LoadingMessage />}>
                <Switch>
                  <Route exact path="/">
                    <AsyncPublications />
                  </Route>
                  <Route path="/events">
                    <AsyncEvents />
                  </Route>
                  <Route path="/codelancer">
                    <AsyncCodeChallenge />
                  </Route>
                  <Route path="/notfound">
                    <AsyncNotFound />
                  </Route>
                  <Redirect to="/notfound" />
                </Switch>
              </Suspense>
            </ScrollToTop>
          </FullHeight>
          <Footer />
        </MuiThemeProvider>
      </Router>
    );
  }

  // TODO automatically determine all pages & order from Route, do not use document.body.style side-effect
  private updateBackgroundColor(pathname: string) {
    const page = pathname.split('/')[1];
    const index = pages.indexOf(page);
    const position = -(index >= 0 ? index : 0) * 100;
    document.body.style.backgroundPositionY = `${position}vh, 0`;
  }
}

export default App;
