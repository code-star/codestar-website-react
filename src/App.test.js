import React from 'react';
import { shallow, mount } from 'enzyme';

import App from './App';

jest.mock('./ScrollToTop', () => 'ScrollToTop');
jest.mock('./Footer/Footer', () => 'Footer');
jest.mock('./Jobs/JobsList', () => {
  return [{ path: 'somePath' }, { path: 'someOtherPath' }];
});
jest.mock('./AsyncComponent/AsyncComponent', () => 'AsyncComponent');
jest.mock('./modules/NavContainer/NavContainer', () => 'NavContainer');

const renderShallow = () => {
  return shallow(<App />);
};

const renderMount = () => {
  return mount(<App />);
};

describe('<App />', () => {
  let wrapper;
  let renderSpy;

  beforeEach(() => {
    renderSpy = jest.spyOn(App.prototype, 'render');
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('Component Lifecycle', () => {
    test('must call render', () => {
      wrapper = renderShallow();
      expect(renderSpy).toHaveBeenCalledTimes(1);
    });
  });

  describe('Instance', () => {
    test('must have an instance of App', () => {
      wrapper = renderMount();
      expect(wrapper.find('App')).toHaveLength(1);
    });
  });

  describe('Snaphot', () => {
    test('default', () => {
      expect(global.renderToJSON(renderShallow())).toMatchSnapshot();
    });
  });
});
