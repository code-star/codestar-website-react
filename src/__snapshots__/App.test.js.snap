// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> Snaphot default 1`] = `
Array [
  <NavContainer
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  />,
  <ScrollToTop>
    <AsyncComponent
      component={[Function]}
      fullHeight={true}
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
      location={
        Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        }
      }
      match={
        Object {
          "isExact": true,
          "params": Object {},
          "path": "/",
          "url": "/",
        }
      }
      staticContext={undefined}
    />
  </ScrollToTop>,
  <Footer />,
]
`;
